sample-programs/erc721/erc721.scm

```diff
@@ -0,0 +1,131 @@
+(define-map accounts-operator
+  ((operator principal) (account principal))
+  ((is-approved bool)))
+
+;; Internals
+
+;; Gets the amount of tokens owned by the specified address.
+(define (balance-of (account principal))
+  (let ((balance
+      (get count 
+        (fetch-entry tokens-count (tuple (owner account))))))
+    (if (eq? balance 'null) 0 balance)))
+
+;; Gets the owner of the specified token ID.
+(define (owner-of (token-id int)) 
+  (get owner 
+    (fetch-entry tokens-owner (tuple (token-id token-id)))))
+
+;; Gets the approved address for a token ID, or zero if no address set (approved method in ERC721)
+(define (is-spender-approved (spender principal) (token-id int))
+  (let ((spender
+      (get spender 
+        (fetch-entry tokens-spender (tuple (token-id token-id))))))
+    (if (eq? spender 'null) 'false 'true)))
+
+;; Tells whether an operator is approved by a given owner (isApprovedForAll method in ERC721)
+(define (is-operator-approved (account principal) (operator principal))
+  (let ((is-approved
+      (get is-approved 
+        (fetch-entry accounts-operator (tuple (operator operator) (account account))))))
+    (if (eq? is-approved 'null) 'false is-approved)))
+
+;; Returns whether the given actor can transfer a given token ID.
+;; To be optimized
+(define (can-transfer (actor principal) (token-id int)) 
+  (or 
+    (eq? actor (owner-of token-id)) 
+    (is-spender-approved actor token-id)
+    (is-operator-approved (owner-of token-id) actor)))
+ 
+;; Internal - Register token
+(define (register-token! (new-owner principal) (token-id int))
+  (let ((current-balance (balance-of new-owner)))
+    (begin
+      (set-entry! tokens-owner 
+        (tuple (token-id token-id))
+        (tuple (owner new-owner))) 
+      (set-entry! tokens-count 
+        (tuple (owner new-owner))
+        (tuple (count (+ 1 current-balance)))) 
+      'true)))
+
+;; Internal - Release token
+(define (release-token! (owner principal) (token-id int))
+  (let ((current-balance (balance-of owner)))
+    (begin
+      (delete-entry! tokens-spender 
+        (tuple (token-id token-id))) 
+      (set-entry! tokens-count 
+        (tuple (owner owner))
+        (tuple (count (- current-balance 1)))) 
+      'true)))
+
+;; Public functions
+
+;; Approves another address to transfer the given token ID (approve method in ERC721)
+;; To be optimized
+(define-public (set-spender-approval (spender principal) (token-id int))
+  (if (eq? spender tx-sender)
+    'false
+    (if (or (eq? tx-sender (owner-of token-id)) 
+            (is-operator-approved tx-sender (owner-of token-id)))
+      (begin
+        (set-entry! tokens-spender 
+          (tuple (token-id token-id))
+          (tuple (spender spender))) 
+      'true)
+      'false)))
+
+;; Sets or unsets the approval of a given operator (setApprovalForAll method in ERC721)
+(define-public (set-operator-approval (operator principal) (is-approved bool))
+  (if (eq? operator tx-sender)
```

Comment:
Is the `operator` argument necessary if it's always going to be `tx-sender`?

Reply:
If you look at the next line, you'll see that we are actually aborting the execution of this method if the `operator` is `tx-sender`

Reply:
Yes, I can see that. But, you haven't answered my question ;) Why bother having `spender` as an argument in the first place if you're always required to pass `tx-sender`?